mojo/stdlib/std/os/process.mojo

```
@@ -0,0 +1,441 @@
+
+        Returns:
+            A `ProcessStatus` for a running process.
+        """
+        return Self()
+
+    fn has_exited(self) -> Bool:
+        """Checks if the process has terminated.
+
+        Returns:
+            True if the process has terminated, either normally or by a signal.
+        """
+        return Bool(self.exit_code) or Bool(self.term_signal)
+
+
+struct Pipe:
+    """Create a pipe for interprocess communication.
+
+    Example usage:
+    ```
+    pipe().write_bytes("TEST".as_bytes())
+    ```
+    """
+
+    var fd_in: Optional[FileDescriptor]
+    """File descriptor for pipe input."""
+    var fd_out: Optional[FileDescriptor]
+    """File descriptor for pipe output."""
+
+    fn __init__(
+        out self,
+        in_close_on_exec: Bool = False,
+        out_close_on_exec: Bool = False,
+    ) raises:
+        """Initializes a new `Pipe`.
+
+        Args:
+            in_close_on_exec: Close the read side of pipe if an `exec` syscall
+              is issued in the process.
+            out_close_on_exec: Close the write side of pipe if an `exec`
+              syscall is issued in the process.
+
+        Raises:
+            Error: If the pipe could not be created or configured.
+        """
+        var pipe_fds = alloc[c_int](2)
+        if pipe(pipe_fds) < 0:
+            pipe_fds.free()
+            raise Error("Failed to create pipe")
+
+        if in_close_on_exec:
+            if not self._set_close_on_exec(pipe_fds[0]):
+                pipe_fds.free()
+                raise Error("Failed to configure input pipe close on exec")
+
+        if out_close_on_exec:
+            if not self._set_close_on_exec(pipe_fds[1]):
+                pipe_fds.free()
+                raise Error("Failed to configure output pipe close on exec")
+
+        self.fd_in = FileDescriptor(Int(pipe_fds[0]))
+        self.fd_out = FileDescriptor(Int(pipe_fds[1]))
+        pipe_fds.free()
+
+    fn __del__(deinit self):
```

I think we have some resource management issues here.  `close()` can fail (e.g. `EINTR`, `EIO`) but the errors are ignored here.  This may be easier to ensure correctness if we initialize `fd_in` and `fd_out` to `None` before any operations that raise in `__init__` for example. Then we could check the return values from `close()` and handle errors appropriately.

---

It's actually OK to swallow errors from close. EBADF is basically structurally impossible, historically everybody except HP-UX would close the fd before returning EINTR, creating inconsistency. Now the open group has said the FD _MUST_ still be open if EINTR is returned but EINPROGRESS should indicate the call was interrupted but the fd was closed.

Sadly this change needs to be adopted by linux. Basically everyone still does the linux semantics.

I don't think we need to care about HP-UX, we can assume the sane linux semantics.

For other IO errors there's nothing we can do by the time we get to this close call. If the user wanted to handle them they would have synched the file.


See:
https://lwn.net/Articles/576478/
https://ewontfix.com/4/

---

@barcharcraz Can you include that last comment inline in the function (no need for a docstring) as a note for future readers? If Linux ever updates its behavior (unlikely) we'll need to fix that, and there's a non-zero chance of another POSIX OS (Redox comes to mind) being written to the spec with the now-correct behavior. Having that information there will save a lot of time.